HDDS-1301. Optimize recursive ozone filesystem apis

lokeshj1703 opened the pull request:
This Jira aims to optimise recursive apis in ozone file system. These are the apis which have a recursive flag which requires an operation to be performed on all the children of the directory. The Jira would add support for recursive apis in Ozone manager in order to reduce the number of rpc calls to Ozone Manager. Also currently these operations are not atomic. This Jira would make all the operations in ozone filesystem atomic.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 32 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 44 | Maven dependency ordering for branch |
| +1 | mvninstall | 1098 | trunk passed |
| +1 | compile | 1102 | trunk passed |
| +1 | checkstyle | 193 | trunk passed |
| -1 | mvnsite | 36 | ozone-manager in trunk failed. |
| +1 | shadedclient | 1127 | branch has no errors when building and testing our client artifacts. |
| -1 | findbugs | 28 | ozone-manager in trunk failed. |
| +1 | javadoc | 162 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for patch |
| -1 | mvninstall | 21 | client in the patch failed. |
| -1 | mvninstall | 21 | ozonefs in the patch failed. |
| +1 | compile | 992 | the patch passed |
| +1 | cc | 992 | the patch passed |
| +1 | javac | 992 | the patch passed |
| +1 | checkstyle | 182 | the patch passed |
| -1 | mvnsite | 31 | ozonefs in the patch failed. |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 696 | patch has no errors when building and testing our client artifacts. |
| -1 | findbugs | 30 | ozonefs in the patch failed. |
| +1 | javadoc | 179 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 78 | common in the patch passed. |
| +1 | unit | 30 | client in the patch passed. |
| -1 | unit | 40 | common in the patch failed. |
| -1 | unit | 54 | ozone-manager in the patch failed. |
| -1 | unit | 34 | ozonefs in the patch failed. |
| +1 | asflicense | 48 | The patch does not generate ASF License warnings. |
| | | 7095 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.om.exceptions.TestResultCodes |
|   | hadoop.ozone.om.ratis.TestOzoneManagerRatisServer |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/718 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  cc  |
| uname | Linux 1b061ac979fe 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 32722d2 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/branch-mvnsite-hadoop-ozone_ozone-manager.txt |
| findbugs | v3.1.0-RC1 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/branch-findbugs-hadoop-ozone_ozone-manager.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/patch-mvninstall-hadoop-ozone_client.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/patch-mvninstall-hadoop-ozone_ozonefs.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/patch-mvnsite-hadoop-ozone_ozonefs.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/patch-findbugs-hadoop-ozone_ozonefs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/patch-unit-hadoop-ozone_common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/patch-unit-hadoop-ozone_ozone-manager.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/artifact/out/patch-unit-hadoop-ozone_ozonefs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/testReport/ |
| Max. process+thread count | 446 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-ozone/client hadoop-ozone/common hadoop-ozone/ozone-manager hadoop-ozone/ozonefs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

bharatviswa504 commented:
Hi @lokeshj1703 
Thanks for the patch.
We want to convert all OM requests in 2 phase for HA. So that in applyTransaction it should be just applied to OM DB. So, that read from OM DB is not required during applyTransaction and also followers can just apply to OM DB.(No additional changes required). And also this helps in future if we want to batch transactions and then commit to OM DB also. Another advantage is all requests will be executed only on the leader OM, if any errors like volume not found, bucket not found etc., will not be written to raft log, as we return the startTransaction with exception.

For more info refer HDDS-1393 and HDDS-1379. If you think any other approach let us know.

mukul1987 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:
```diff
@@ -1546,6 +1559,358 @@ public OmKeyInfo lookupFile(OmKeyArgs args) throws IOException {
+   *                     if recursive flag is false and directory is not empty
+   */
+  public void delete(OmKeyArgs args, boolean recursive) throws IOException {
+    // TODO: add recursive option
```
This has been implemented

lokeshj1703 commented on hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneFileSystem.java:
```diff
@@ -319,337 +260,45 @@ boolean processKey(String key) throws IOException {
       }
-      return null;
-    }
+    } while (tmpStatus.size() == numEntries);
```
This needs to be tmpStatus because we are checking the size of the list obtained in the last listStatus call. If the number of entries are less than the number of entries expected then there are no more entries left to list.

lokeshj1703 commented:
@mukul1987  sorry I actually worked on the last patch itself and therefore changes made in the latest commit will not be visible explicitly. But the latest commit addresses your review comments. Thanks for reviewing the patch!

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 57 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 70 | Maven dependency ordering for branch |
| +1 | mvninstall | 1077 | trunk passed |
| +1 | compile | 1063 | trunk passed |
| +1 | checkstyle | 138 | trunk passed |
| -1 | mvnsite | 52 | client in trunk failed. |
| +1 | shadedclient | 1127 | branch has no errors when building and testing our client artifacts. |
| -1 | findbugs | 33 | client in trunk failed. |
| +1 | javadoc | 186 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 22 | Maven dependency ordering for patch |
| -1 | mvninstall | 19 | client in the patch failed. |
| -1 | mvninstall | 18 | ozonefs in the patch failed. |
| +1 | compile | 1035 | the patch passed |
| +1 | cc | 1035 | the patch passed |
| +1 | javac | 1035 | the patch passed |
| +1 | checkstyle | 151 | the patch passed |
| -1 | mvnsite | 35 | ozonefs in the patch failed. |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 684 | patch has no errors when building and testing our client artifacts. |
| -1 | findbugs | 33 | ozonefs in the patch failed. |
| +1 | javadoc | 181 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 95 | common in the patch passed. |
| +1 | unit | 36 | client in the patch passed. |
| -1 | unit | 44 | common in the patch failed. |
| -1 | unit | 75 | ozone-manager in the patch failed. |
| -1 | unit | 32 | ozonefs in the patch failed. |
| +1 | asflicense | 44 | The patch does not generate ASF License warnings. |
| | | 7183 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.om.exceptions.TestResultCodes |
|   | hadoop.ozone.om.ratis.TestOzoneManagerRatisServer |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/718 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  cc  |
| uname | Linux 18bc982ec5df 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 13907d8 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/branch-mvnsite-hadoop-ozone_client.txt |
| findbugs | v3.1.0-RC1 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/branch-findbugs-hadoop-ozone_client.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/patch-mvninstall-hadoop-ozone_client.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/patch-mvninstall-hadoop-ozone_ozonefs.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/patch-mvnsite-hadoop-ozone_ozonefs.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/patch-findbugs-hadoop-ozone_ozonefs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/patch-unit-hadoop-ozone_common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/patch-unit-hadoop-ozone_ozone-manager.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/artifact/out/patch-unit-hadoop-ozone_ozonefs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/testReport/ |
| Max. process+thread count | 414 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-ozone/client hadoop-ozone/common hadoop-ozone/ozone-manager hadoop-ozone/ozonefs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-718/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.
